### PR TITLE
[SPARK-44983][SQL] Convert binary to string by `to_char` for the formats: `hex`, `base64`, `utf-8`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to re-use the `Hex`, `Base64` and `Decode` expressions in the `ToCharacter` (the `to_char`/`to_varchar` functions) when the `format` parameter is one of `hex`, `base64` and `utf-8`.

### Why are the changes needed?
To make the migration to Spark SQL easier from the systems like:
- Snowflake: https://docs.snowflake.com/en/sql-reference/functions/to_char
- SAP SQL Anywhere: https://help.sap.com/docs/SAP_SQL_Anywhere/93079d4ba8e44920ae63ffb4def91f5b/81fe51196ce21014b9c6cf43b2666698.html
- Oracle: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/TO_CHAR-number.html#GUID-00DA076D-2468-41AB-A3AC-CC78DBA0D9CB
- Vertica: https://www.vertica.com/docs/9.3.x/HTML/Content/Authoring/SQLReferenceManual/Functions/Formatting/TO_CHAR.htm

### Does this PR introduce _any_ user-facing change?
No. This PR extends existing API. It might be considered as an user-facing change only if user's code depends on errors in the case of wrong formats.

### How was this patch tested?
By running new examples:
```
$ build/sbt "sql/test:testOnly org.apache.spark.sql.expressions.ExpressionInfoSuite"
```
and new tests:
```
$ build/sbt "test:testOnly *.StringFunctionsSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala [hunk 1]
```diff
@@ -854,28 +854,53 @@ class StringFunctionsSuite extends QueryTest with SharedSparkSession {
...
+      checkAnswer(df2.selectExpr(s"$funcName(input, 'base64')"), Seq(Row("ZA==")))
+      checkAnswer(df2.select(func(col("input"), lit("base64"))), Seq(Row("ZA==")))
+      checkAnswer(df2.selectExpr(s"$funcName(input, 'utf-8')"), Seq(Row("d")))
+      checkAnswer(df2.select(func(col("input"), lit("utf-8"))), Seq(Row("d")))
+
+      checkError(
+        exception = intercept[AnalysisException] {
+          df2.select(func(col("input"), col("format"))).collect()
+        },
+        errorClass = "_LEGACY_ERROR_TEMP_1100",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala:878]: TODO: Open an JIRA to assign proper name for the error class.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala [hunk 2]
```diff
@@ -232,10 +242,20 @@ object ToCharacterBuilder extends ExpressionBuilder {
   override def build(funcName: String, expressions: Seq[Expression]): Expression = {
     val numArgs = expressions.length
     if (expressions.length == 2) {
-      val inputExpr = expressions.head
+      val (inputExpr, format) = (expressions(0), expressions(1))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala:245]: this is a bit weird, we can just write 2 lines
```
val inputExpr = expressions(0)
val format = expressions(1)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -157,6 +157,15 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
         "functionName" -> toSQLId("format_string")))
   }
 
+  def binaryFormatError(funcName: String, invalidFormat: String): Throwable = {
+    new AnalysisException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:161]: My feeling is that, if the error is only thrown in one place, we don't need to add a method here, just `throw new AnalysisException...` in the caller side.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:161]: One of purposes of gathering exceptions to `Query*Errors` to don't depend on forming exceptions: errors classes, quoting, context and so on. The caller has to provide only valuable info and don't worry about any technical stuff.

Don't think that this PR is right place to begin don't do that.

File: sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala:878]: https://issues.apache.org/jira/browse/SPARK-44987

**Comment by MaxGekk** [comment]
All GAs passed. Merging to master.
Thank you, @HyukjinKwon @cloud-fan for review.

File: sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala:878]: Here is the PR https://github.com/apache/spark/pull/42737

**Pull request closed without merging**